In [1]:
import pandas as pd
import numpy as np
import pickle

In [2]:
f = open("./final_data.p", "rb")
data = pickle.load(f)

In [3]:
data.head()

fips_code              county  total_2008  dem_2008  gop_2008  oth_2008  \
0      26041        Delta County     19064.0    9974.0    8763.0     327.0   
1      48295     Lipscomb County      1256.0     155.0    1093.0       8.0   
2       1127       Walker County     28652.0    7420.0   20722.0     510.0   
3      48389       Reeves County      3077.0    1606.0    1445.0      26.0   
4      56017  Hot Springs County      2546.0     619.0    1834.0      93.0   

   total_2012  dem_2012  gop_2012  oth_2012          ...           \
0     18043.0    8330.0    9533.0     180.0          ...            
1      1168.0     119.0    1044.0       5.0          ...            
2     28497.0    6551.0   21633.0     313.0          ...            
3      2867.0    1649.0    1185.0      33.0          ...            
4      2495.0     523.0    1894.0      78.0          ...            

   Family Income 35-40K  Family Income 40-45K  Family Income 45-50K  \
0                 453.0                 572.0                 528.0   
1                  20.0                  29.0                  33.0   
2                1041.0                 919.0                 987.0   
3                  96.0                 119.0                  75.0   
4                  50.0                  69.0                  84.0   

   Family Income 50-60K  Family Income 60-75K  Family Income 75-100K  \
0                1217.0                1261.0                 1697.0   
1                  79.0                 123.0                  116.0   
2                1570.0                2209.0                 1819.0   
3                 213.0                 320.0                  350.0   
4                 122.0                 234.0                  112.0   

   Family Income 100-125K  Family Income 125-150K  Family Income 150-200K  \
0                   623.0                   401.0                   234.0   
1                    91.0                    54.0                    56.0   
2                  1210.0                   745.0                   437.0   
3                   188.0                    69.0                    64.0   
4                   148.0                    89.0                    36.0   

   Family Income > 200K  
0                  96.0  
1                  46.0  
2                 339.0  
3                  79.0  
4                  58.0  

[5 rows x 134 columns]

In [4]:
X = data.ix[:, 14:]

In [5]:
X.head()

# of Housing Units  # White  # White Male  # White Female  # Black  \
0              3040.0  34754.0       17231.0         17523.0    121.0   
1               478.0   2980.0        1577.0          1403.0     17.0   
2              2700.0  60224.0       29335.0         30889.0   3764.0   
3               556.0  11899.0        6966.0          4933.0    750.0   
4               326.0   4661.0        2298.0          2363.0     12.0   

   # Black Male  # Black Female  # White Not Hispanic  \
0          60.0            61.0               34581.0   
1           8.0             9.0                2361.0   
2        1781.0          1983.0               59498.0   
3         578.0           172.0                2677.0   
4           1.0            11.0                4535.0   

   # White Not Hispanis Male  # White Not Hispanis Female  \
0                    17137.0                      17444.0   
1                     1194.0                       1167.0   
2                    28884.0                      30614.0   
3                     1795.0                        882.0   
4                     2247.0                       2288.0   

           ...           Family Income 35-40K  Family Income 40-45K  \
0          ...                          453.0                 572.0   
1          ...                           20.0                  29.0   
2          ...                         1041.0                 919.0   
3          ...                           96.0                 119.0   
4          ...                           50.0                  69.0   

   Family Income 45-50K  Family Income 50-60K  Family Income 60-75K  \
0                 528.0                1217.0                1261.0   
1                  33.0                  79.0                 123.0   
2                 987.0                1570.0                2209.0   
3                  75.0                 213.0                 320.0   
4                  84.0                 122.0                 234.0   

   Family Income 75-100K  Family Income 100-125K  Family Income 125-150K  \
0                 1697.0                   623.0                   401.0   
1                  116.0                    91.0                    54.0   
2                 1819.0                  1210.0                   745.0   
3                  350.0                   188.0                    69.0   
4                  112.0                   148.0                    89.0   

   Family Income 150-200K  Family Income > 200K  
0                   234.0                  96.0  
1                    56.0                  46.0  
2                   437.0                 339.0  
3                    64.0                  79.0  
4                    36.0                  58.0  

[5 rows x 120 columns]

In [6]:
del X["Population"]

In [7]:
X = X.apply(lambda x: x / data["Population"])

In [8]:
X.head()

# of Housing Units   # White  # White Male  # White Female   # Black  \
0            0.082517  0.943351      0.467713        0.475639  0.003284   
1            0.139848  0.871855      0.461381        0.410474  0.004974   
2            0.040779  0.909577      0.443053        0.466524  0.056849   
3            0.039777  0.851266      0.498355        0.352912  0.053656   
4            0.067551  0.965810      0.476171        0.489639  0.002487   

   # Black Male  # Black Female  # White Not Hispanic  \
0      0.001629        0.001656              0.938655   
1      0.002341        0.002633              0.690755   
2      0.026899        0.029950              0.898612   
3      0.041351        0.012305              0.191515   
4      0.000207        0.002279              0.939702   

   # White Not Hispanis Male  # White Not Hispanis Female  \
0                   0.465161                     0.473494   
1                   0.349327                     0.341428   
2                   0.436242                     0.462370   
3                   0.128416                     0.063099   
4                   0.465603                     0.474099   

           ...           Family Income 35-40K  Family Income 40-45K  \
0          ...                       0.012296              0.015526   
1          ...                       0.005851              0.008484   
2          ...                       0.015722              0.013880   
3          ...                       0.006868              0.008513   
4          ...                       0.010361              0.014298   

   Family Income 45-50K  Family Income 50-60K  Family Income 60-75K  \
0              0.014332              0.033034              0.034228   
1              0.009655              0.023113              0.035986   
2              0.014907              0.023712              0.033363   
3              0.005366              0.015238              0.022893   
4              0.017406              0.025280              0.048487   

   Family Income 75-100K  Family Income 100-125K  Family Income 125-150K  \
0               0.046063                0.016911                0.010885   
1               0.033938                0.026624                0.015799   
2               0.027473                0.018275                0.011252   
3               0.025039                0.013450                0.004936   
4               0.023208                0.030667                0.018442   

   Family Income 150-200K  Family Income > 200K  
0                0.006352              0.002606  
1                0.016384              0.013458  
2                0.006600              0.005120  
3                0.004579              0.005652  
4                0.007460              0.012018  

[5 rows x 119 columns]

In [9]:
X["# Female"].describe()

count    2628.000000
mean        0.500386
std         0.022183
min         0.302915
25%         0.496416
50%         0.504755
75%         0.511384
max         0.571966
Name: # Female, dtype: float64

In [10]:
voter_turn_out = data["total_2016"] - data["total_2012"]

In [11]:
from sklearn.linear_model import LinearRegression

In [12]:
X = X * 100

In [13]:
X.shape

(2628, 119)

In [14]:
linear = LinearRegression()

In [15]:
linear.fit(X, voter_turn_out)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [16]:
for name, val in sorted(zip(X.columns, linear.coef_), key=lambda x: x[1], reverse=True):
    print((name, val))

('Median Age Male', 35273.860299653905)
('Median Age Female', 14903.935966891951)
('Rail Road', 8814.3766146147937)
('Family Income > 200K', 3049.9008713594717)
('Moved From Abroad White', 2675.020037522846)
('HS or equal', 2658.6651162152498)
('Total Moving White', 2585.6870454585942)
('Other State-Bachelor', 2495.9051199867345)
('Graduate', 2457.4716516865096)
('Family Income 125-150K', 2284.1992306021039)
('Family Income 150-200K', 2250.1983121520148)
('No Vehicle Owned', 2229.9841806845379)
('In State-Less than HS', 2134.8441168842924)
('Family Income 100-125K', 2093.645998593141)
('Other State-Less than HS', 1725.9416588164067)
('In State-Bachelor', 1636.9291933644913)
('Family Income 75-100K', 1366.1599178436029)
('Public Transportation', 1046.0014325030434)
('Income < 10K', 920.03628001649167)
('Family Income 60-75K', 844.4556938609162)
('Other State-Some College', 818.02963755389555)
('In State-Some College', 812.11617145943251)
('# Male', 709.39649665300635)
('Total Reporing 0

In [17]:
from sklearn.linear_model import LassoCV

In [18]:
lasso = LassoCV()

In [19]:
lasso.fit(X, voter_turn_out)

/opt/conda/lib/python3.5/site-packages/sklearn/linear_model/coordinate_descent.py:479: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)


LassoCV(alphas=None, copy_X=True, cv=None, eps=0.001, fit_intercept=True,
    max_iter=1000, n_alphas=100, n_jobs=1, normalize=False, positive=False,
    precompute='auto', random_state=None, selection='cyclic', tol=0.0001,
    verbose=False)

In [20]:
keep_vars = []

for name, val in zip(X.columns, lasso.coef_):    
    if val != 0:
        keep_vars.append(name)

In [21]:
linear = LinearRegression()

In [23]:
linear.fit(X[keep_vars], voter_turn_out)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [24]:
for name, val in sorted(zip(X.columns, linear.coef_), key=lambda x: x[1], reverse=True):
    if val != 0:
        print((name, val))

('In State-Some College', 9872.7561185049144)
('Other State-Less than HS', 2058.9858569503672)
('# Hispanic Male', 1288.4624821200146)
('In State-HS or equal', 845.68878360072529)
('Median Age Male', 496.55432424656192)
('HS or equal', 401.19128869224431)
('# White', 363.68263718331627)
('# Female', 359.28511922859684)
('Other State-Some College', 352.17363970568852)
('# White Female', 293.88876481382579)
('Other State-HS or equal', 274.16132283847315)
('Graduate', 246.65760644245992)
('Bachelor', 198.10990608368564)
('Some College-AA', 189.65179171791351)
('# White Not Hispanis Female', 146.54675332255715)
('# of Housing Units', 112.42024521513521)
('Less Than HS', 91.94225346885284)
('Total Born In State education', 89.712109861520091)
('Other State-Bachelor', 79.907528031328454)
('Education Count', 35.320520899747613)
('Citizen: born in US', 17.899793709838555)
('Separated', 15.14643544401876)
('In State-Graduate', 1.8623403333036568)
('Never Married', 0.37454060582691456)
('Current

How some of these can be interpretted.

('# Female', 359.28511922859684)

After adjusting for other predictors, each 1% increase in the number of females in a county is associated with an increase in the mean voter turn out from 2012 to 2016 by 359 voters.

In [ ]:
diff_voting_2012 = data["dem_2012"] - data["gop_2012"]
diff_voting_2016 = data["dem_2016"] - data["gop_2016"]

y = diff_voting_2016 - diff_voting_2012

In [ ]:
ratio_2012 = data["dem_2012"] / data["total_2012"] - 0.5
ratio_2016 = data["dem_2016"] / data["total_2016"] - 0.5

ratio_change = diff_voting_2016 - diff_voting_2012